third_party/tonic/dart_weak_persistent_value.cc

```diff
@@ -0,0 +1,79 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "tonic/dart_weak_persistent_value.h"
+
+#include "tonic/dart_state.h"
+#include "tonic/scopes/dart_isolate_scope.h"
+
+namespace tonic {
+
+DartWeakPersistentValue::DartWeakPersistentValue() : value_(nullptr) {}
+
+DartWeakPersistentValue::DartWeakPersistentValue(
+    DartWeakPersistentValue&& other)
+    : dart_state_(other.dart_state_), value_(other.value_) {
+  other.dart_state_.reset();
+  other.value_ = nullptr;
+}
+
+DartWeakPersistentValue::DartWeakPersistentValue(
+    DartState* dart_state,
+    Dart_Handle object,
+    void* peer,
+    intptr_t external_allocation_size,
+    Dart_HandleFinalizer callback)
+    : value_(nullptr) {
+  Set(dart_state, object, peer, external_allocation_size, callback);
+}
+
+DartWeakPersistentValue::~DartWeakPersistentValue() {
+  Clear();
+}
+
+void DartWeakPersistentValue::Set(DartState* dart_state,
+                                  Dart_Handle object,
+                                  void* peer,
+                                  intptr_t external_allocation_size,
+                                  Dart_HandleFinalizer callback) {
+  TONIC_DCHECK(is_empty());
+  dart_state_ = dart_state->GetWeakPtr();
+  value_ = Dart_NewWeakPersistentHandle(object, peer, external_allocation_size,
+                                        callback);
+}
+
+void DartWeakPersistentValue::Clear() {
+  if (!value_) {
+    return;
+  }
+
+  auto dart_state = dart_state_.lock();
+  if (!dart_state) {
+    return;
+  }
+
+  if (Dart_CurrentIsolateGroup()) {
+    Dart_DeleteWeakPersistentHandle(value_);
+  } else if (!dart_state->IsShuttingDown()) {
+    DartIsolateScope scope(dart_state->isolate());
```

Comment:
Not all use sites maintain this invariant. For example: https://gist.github.com/dcharkes/1ac1c9eff900bb58bc854dc27ab433e0.

I will investigate changing the invariant later.

Reply:
Changing this invariant seems nontrivial. For example these destructors can get called on some task processing thread that processes all kinds of tasks. Some specific task (processing image descriptors) can reduce the refcount to 0 which causes deletion of the WPH.

However, not enforcing this variant might create the situation that a second thread tries to enter an isolate.
The situation that a thread entering an isolate while the isolate group is shutting down should be prevented by the check on `DartState::is_shutting_down_` the lines above. (Though, maybe there are cases in which this PR does not cover in which `DartState` should be updated.)

> We have not done this before. Why is this now necessary?

Because before WPHs were auto deleting. Now we need to delete them manually (unless the isolate group is shutting down). In order to delete a WPH, we need to be in an isolate group.

I have observed a flake locally, and managed to catch it in GDB once, that hits a problem with this. Either (1) DartState::is_shutting_down_ is not set correctly, or a second thread is trying to enter an isolate. Repro command: `~/flt/engine/flutter/packages/flutter$ for i in `seq 100`; do ../../bin/flutter test --track-widget-creation --null-assertions --no-sound-null-safety --local-engine host_debug --local-engine-src-path /usr/local/google/home/dacoharkes/flt/engine/src test/cupertino/ ; done`.

This PR needs more work to address this.

Reply:
Addressed flaky crash.